### ISSUE TITLE
Rule blocked: CORERULES-9617 - validation fails when operation references a domain that does not exist

### ISSUE THREAD

**Issue by ASL-rmarshall** [cdisc-org/cdisc-rules-engine#917]
- A .txt file with the JSON containing the full request: [Request.txt](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBNXJVQnc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--a6e72e838d3cf44ad2f1469fb02696f58f3ea101/Request.txt)

**Links to related JIRA Tickets**
- https://jira.cdisc.org/projects/CORERULES/issues/CORERULES-9617

**Rule Information**
- **Standard**: USDM
- **Rule ID**: DDF00074
- **Rule Description**: If the intervention model indicates a single group design then only one intervention is expected. In all other cases more interventions are expected.

**Describe the bug**
This rule is checking whether the number of records describing study interventions (from the StudyIntervention dataset) corresponds with the value of the interventionModel code or decode.  The rule works correctly when there is a StudyIntervention dataset, but it fails with the following error when there is no StudyIntervention dataset:

> argument of type 'NoneType' is not iterable.

The error occurs because the rule includes an operation that references the "STUDYINTERVENTION" domain but a corresponding dataset does not exist, which would be possible if the original USDM JSON file did not contain any instances of the `StudyIntervention` class.

The following is a stacktrace for the error:

> Error: argument of type 'NoneType' is not iterable, line 252 in get_dataset_name_from_details
Trace(1): <FrameSummary file C:\CDISC\cdisc-rules-engine\cdisc_rules_engine\rules_engine.py, line 168 in validate_single_rule>
Trace(2): <FrameSummary file C:\CDISC\cdisc-rules-engine\cdisc_rules_engine\rules_engine.py, line 301 in validate_rule>
Trace(3): <FrameSummary file C:\CDISC\cdisc-rules-engine\cdisc_rules_engine\rules_engine.py, line 340 in execute_rule>
Trace(4): <FrameSummary file C:\CDISC\cdisc-rules-engine\cdisc_rules_engine\utilities\rule_processor.py, line 284 in perform_rule_operations>
Trace(5): <FrameSummary file C:\CDISC\cdisc-rules-engine\cdisc_rules_engine\utilities\rule_processor.py, line 329 in _execute_operation>
Trace(6): <FrameSummary file C:\CDISC\cdisc-rules-engine\cdisc_rules_engine\utilities\utils.py, line 252 in get_dataset_name_from_details>

It shows that validation is falling over when a `None` is passed into `get_dataset_name_from_details`, which expects a dictionary.

**Error returned from Rule Engine**
```json
{
    "STUDYDESIGN": [
        {
            "executionStatus": "execution_error",
            "dataset": "StudyDesign.xpt",
            "domain": "STUDYDESIGN",
            "variables": [],
            "message": "rule execution error",
            "errors": [
                {
                    "dataset": "StudyDesign.xpt",
                    "error": "An unknown exception has occurred",
                    "message": "argument of type 'NoneType' is not iterable"
                }
            ]
        }
    ]
}
```

**Expected behavior**
Instead of failing with an error, the validation should just skip any operation that references a domain that does not exist.  This could be achieved by inserting the following between the assignment of `domain_details` and the call to `get_dataset_name_from_details` in `_execute_operation` (i.e., at line 329 in `rule_processor.py`):
```python
            if domain_details is None:
                return dataset
```
In this case, if the referenced domain does not exist, the call to the operation just returns the original dataset (without the operation-specific `id` variable).  The presence/absence of the result of the operation can then be checked in the rule condition (as shown in rule DDF00074 in the production environment or the attached Request.txt file).

**Comment by ASL-rmarshall** [comment]
Test data is available in Excel format in the unitTesting SharePoint under USDM > DDF00074.  There are two negative test cases and one positive test case.  It is only the negative test case 02 (where there is no StudyIntervention dataset) that fails as described above.

**Comment by ASL-rmarshall** [comment]
Functionality has been changed (#970) so that the rule is now skipped (instead of reporting an error) if an operation references a dataset/domain that does not exist.

It might be helpful to have an optional parameter that allows the rule to continue execution if a dataset/domain is referenced by an operation but does not exist.

For example, a `domain_required` boolean flag could be used (with a default of `True`). If `True` and the referenced dataset/domain does not exist, the existing error handling would work as it currently does, with the rule being skipped. If `False`, execution could continue without the operation's id variable being added to the check dataset.  The existence of the id variable could the be checked using operators such as `exists`.

**Comment by ASL-rmarshall** [comment]
In general, it would be easier to use JSONata to define a check where there may be no instances of a checked class.  Changing the way that tabular checking deals with missing datasets would probably be too big a change.
